packages/synapse-react-client/src/components/DynamicForm/DynamicForm.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { useEffect, useState } from 'react'
+import Form from '@rjsf/mui'
+import validator from '@rjsf/validator-ajv8'
+import { displayToast } from '../ToastMessage'
+import { Box, Button, Typography } from '@mui/material'
+import { SkeletonParagraph } from '../Skeleton'
+
+export type DynamicFormProps = {
+  schemaUrl: string
+  uiSchemaUrl: string
+  postUrl: string
+  mutateFormDataBeforePost?: (formData: any) => any // allow caller to embed formdata into custom request body object
+  onCancel?: () => void
+}
+
+function DynamicForm(props: DynamicFormProps) {
+  const {
+    schemaUrl,
+    uiSchemaUrl,
+    postUrl,
+    mutateFormDataBeforePost,
+    onCancel,
+  } = props
+  const [schema, setSchema] = useState(null)
+  const [uiSchema, setUiSchema] = useState(null)
+  const [formData, setFormData] = useState({})
+  const [isSubmitting, setIsSubmitting] = useState(false)
+
+  useEffect(() => {
+    // Fetch the form schema
+    const fetchSchema = async () => {
+      try {
+        const schemaResponse = await fetch(schemaUrl)
+        const schemaData = await schemaResponse.json()
+        setSchema(schemaData)
+      } catch (error) {
+        displayToast(`Error fetching Schema: ${error}`, 'danger')
+      }
+    }
+
+    // Fetch the UI schema
+    const fetchUiSchema = async () => {
+      try {
+        const uiSchemaResponse = await fetch(uiSchemaUrl)
+        const uiSchemaData = await uiSchemaResponse.json()
+        setUiSchema(uiSchemaData)
+      } catch (error) {
+        displayToast(`Error fetching UI Schema: ${error}`, 'danger')
+      }
+    }
+
+    fetchSchema()
+    fetchUiSchema()
+  }, [schemaUrl, uiSchemaUrl])
+
+  const handleSubmit = async (formData: any) => {
+    setIsSubmitting(true)
+    const formDataToSubmit = mutateFormDataBeforePost
+      ? mutateFormDataBeforePost(formData)
+      : formData
+    try {
+      const response = await fetch(postUrl, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+          Accept: 'application/json',
+        },
+        body: JSON.stringify(formDataToSubmit),
+      })
+
+      if (response.ok) {
+        displayToast('Form submitted successfully!', 'success')
+      } else {
+        const responseText = await response.text()
+        displayToast(`Form submission failed: ${responseText}`, 'danger')
+      }
+    } catch (error) {
+      displayToast(`Form submission failed: ${JSON.stringify(error)}`, 'danger')
+    } finally {
+      setIsSubmitting(false)
+    }
+  }
+
+  if (!schema || !uiSchema) {
```

Comment:
Use `isLoading` returned by `useQuery`.

Is it true that there will always be a `uiSchema`? It's optional in RJSF

Reply:
For NF cases so far, UI schema is absolutely necessary.  If we encounter a case where it is not needed we can update